openmls/src/ciphersuite/hash_ref.rs

```diff
@@ -58,45 +78,73 @@ pub type KeyPackageRef = HashReference;
 /// This value uniquely identifies a proposal.
 pub type ProposalRef = HashReference;
 
+#[derive(TlsSerialize, TlsSize)]
+struct HashReferenceInput<'a> {
+    label: TlsSliceU8<'a, u8>,
+    value: VLBytes,
+}
+
+/// Compute a new [`ProposalRef`] value for a `value`.
+pub fn make_proposal_ref(
+    value: &[u8],
+    ciphersuite: Ciphersuite,
+    backend: &impl OpenMlsCrypto,
+) -> Result<ProposalRef, CryptoError> {
+    HashReference::new(value, ciphersuite, backend, PROPOSAL_REF_LABEL)
+}
+
+/// Compute a new [`KeyPackageRef`] value for a `value`.
+pub fn make_key_package_ref(
+    value: &[u8],
+    ciphersuite: Ciphersuite,
+    backend: &impl OpenMlsCrypto,
+) -> Result<KeyPackageRef, CryptoError> {
+    HashReference::new(value, ciphersuite, backend, KEY_PACKAGE_REF_LABEL)
+}
+
 impl HashReference {
     /// Compute a new [`HashReference`] value for a `value`.
     pub fn new(
         value: &[u8],
         ciphersuite: Ciphersuite,
         backend: &impl OpenMlsCrypto,
+        label: &[u8],
     ) -> Result<Self, CryptoError> {
-        let okm = backend.hkdf_expand(
-            ciphersuite.hash_algorithm(),
-            &backend.hkdf_extract(ciphersuite.hash_algorithm(), &[], value)?,
-            LABEL,
-            VALUE_LEN,
-        )?;
-        let value: Value = okm.try_into().map_err(|_| CryptoError::InvalidLength)?;
-        Ok(Self { value })
+        let input = HashReferenceInput {
+            label: TlsSliceU8(label),
+            value: VLBytes::new(value.to_vec()),
+        };
+        let payload = input
+            .tls_serialize_detached()
+            .map_err(|_| CryptoError::TlsSerializationError)?;
+        let value = backend.hash(ciphersuite.hash_algorithm(), &payload)?;
+        Ok(Self {
+            value: Value(VLBytes::new(value)),
+        })
     }
 
     /// Get a reference to the hash reference's value.
-    pub fn value(&self) -> &[u8; 16] {
-        &self.value
+    pub fn value(&self) -> &[u8] {
```

Comment:
Now that values are variable length, should we remove `value` entirely?

Reply:
I was just wondering whether it made sense to have both `fn value()` and `fn as_slice()` since they behave identically now (we can no longer return a fixed size array reference). We could switch to only using `as_slice`.

Reply:
Fair point. Let's get rid of `value` then.